### ISSUE TITLE
wasmer cli framebuffer crashes when file is opened and written to multiple times

### ISSUE THREAD

**Issue by richardanaya** [wasmerio/wasmer#1276]
I wrote some code that fails when I repeated write to the framebuffer. It works on the first time or if I only write to the frame buffer once.

Working:

```rust
pub fn _start() {
    libw::write_text("/_wasmer/dev/fb0/virtual_size","400x400").unwrap();
    let len = 400*400*4;
    let mut pixels = vec![0 as u8;len];

    for i in 0..len {
            pixels[i] = (libw::random_number()*255.0) as u8;
     }
    
    libw::write_bytes("/_wasmer/dev/fb0/fb",&pixels).unwrap();
    libw::write_text("/_wasmer/dev/fb0/draw","0").unwrap();
    loop {
        libw::sleep(1000);
    }
}
```

Failing;

```rust
pub fn _start() {
    libw::write_text("/_wasmer/dev/fb0/virtual_size","400x400").unwrap();
    let len = 400*400*4;
    let mut pixels = vec![0 as u8;len];

    loop {
        for i in 0..len {
            pixels[i] = (libw::random_number()*255.0) as u8;
        }
    
        libw::write_bytes("/_wasmer/dev/fb0/fb",&pixels).unwrap();
        libw::write_text("/_wasmer/dev/fb0/draw","0").unwrap();
        libw::sleep(1000);
    }
}

```

**Comment by richardanaya** [comment]
[working.wasm.txt](https://github.com/wasmerio/wasmer/files/4303501/working.wasm.txt)

**Comment by richardanaya** [comment]
[fails.wasm.txt](https://github.com/wasmerio/wasmer/files/4303502/fails.wasm.txt)

**Comment by richardanaya** [comment]
this appears to apply to /_wasmer/dev/fb0/input too

**Comment by MarkMcCaskey** [comment]
Thanks for the reports in regards to the framebuffer!

It looks like in this case, it should be fixable by using `pwrite` or `seek` after you write.  The framebuffer is emulating a real file so normal calls to `read` and `write` will advance its cursor, so writing the full size will place the cursor at the end of the file causing further writes to fail.

Can you try adding an absolute seek to position 0 or using `pwrite` instead and let me know if that works?

**Comment by MarkMcCaskey** [comment]
Perhaps the behavior is not intuitive and should be documented better somewhere!

**Comment by MarkMcCaskey** [comment]
In regards to the crash, I believe that's due to the `unwrap`!  the write calls should be returning an error indicating that the end of the file has been reached and that no more writes can be done.  If that's not the case, let me know and I'll investigate further on the Wasmer side!

**Comment by richardanaya** [comment]
Forgive my ignorance as i'm not a POSIX expert :) every iteration of the loop I "open" the file, is open supposed to reset the seek position to zero?

**Comment by MarkMcCaskey** [comment]
That's correct!  That is the correct behavior and that is what usually happens.  However the current implementation is a bit hacky and doesn't support that access pattern because the file descriptors are actually passed in before the WASI program starts and so can't actually be reopened once they're closed and open just consults a separate table when "open" is called.  Unfortunately the experimental-io-devices code still has some prototype quality code in it (this bug being the prime example of that...).

This is something I'd like to fix with #1219, but for now `pwrite` or `seek` should do the trick!  `path_open` should be much slower here (compared to `seek` or using `pwrite`) because it has to do a lot more work -- so I'd strongly recommend using `seek` or `pwrite` even if this did work!
